Sources/Jobs/JobInstance.swift

```diff
@@ -0,0 +1,77 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Hummingbird server framework project
+//
+// Copyright (c) 2024 the Hummingbird authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See hummingbird/CONTRIBUTORS.txt for the list of Hummingbird authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+
+/// Protocol for a Job
+protocol JobInstanceProtocol: Sendable {
+    /// Parameters job requries
+    associatedtype Parameters: Codable & Sendable
+    /// Job Type identifier
+    var id: JobIdentifier<Parameters> { get }
+    /// Maximum number of times a job will be retried before being classed as failed
+    var maxRetryCount: Int { get }
+    /// Time job was queued
+    var queuedAt: Date { get }
+    /// Function to execute the job
+    func execute(context: JobContext) async throws
+}
+
+extension JobInstanceProtocol {
+    /// name of job type
+    public var name: String {
```

Comment:
Should the JobName move out of Parameters? Right now, only one job can be registered at a time.

Reply:
You can register multiple jobs to a queue handler. If you can't that's a bug. 

The JobParameters holds both the job parameters and the job identifier. The job name is a string version of the identifier that's used in encoding and decoding.

Reply:
Makes sense.

We can already do the the following today

```swift
queue.registerJob(
            parameters: EmailParameters.self,
            maxRetryCount: 10)
        { (parameters: EmailParameters, _) in
            do {
                try await emailJob.send(emailParameters: parameters)
            } catch {
                logger.info("Failed to send semail: \(error.localizedDescription)", metadata: [
                    "error": "\(String(reflecting: error))",
                ])
                throw error
            }
        }

queue.registerJob(
            parameters: StatsParameters.self,
            maxRetryCount: 5)
        { (parameters: StatsParameters, _) in
            do {
                try await statsJob.compute(date: Date.now)
            } catch {
                logger.info("Failed to compute stats: \(error.localizedDescription)", metadata: [
                    "error": "\(String(reflecting: error))",
                ])
                throw error
            }
        }

queue.registerJob(
            parameters: AnotherParameters.self,
            maxRetryCount: 4)
        { (parameters: AnotherParameters, _) in
        }
```

I was asking if the interface could change were the we pass the name of the job at the time of registration instead.

```swift
queue.registerJob(
            parameters: AnotherParameters.self,
            jobName: "SomeJobName"
            maxRetryCount: 4)
        { (parameters: AnotherParameters, _) in
        }
```

Reply:
Can we discuss this in another issue? My changes here aren't really related.

Reply:
I think this thread can be closed too as the ticket was resolved.